presto-main/src/main/java/com/facebook/presto/operator/OptimizedPartitionedOutputOperator.java

```diff
@@ -0,0 +1,793 @@
+
+        private int bufferedRowCount;
+        private boolean bufferFull;
+
+        PartitionData(int partition, int channelCount, int capacity, AtomicLong pagesAdded, AtomicLong rowsAdded, PagesSerde serde, Lifespan lifespan)
+        {
+            this.partition = partition;
+            this.channelCount = channelCount;
+            this.capacity = capacity;
+            this.pagesAdded = requireNonNull(pagesAdded, "pagesAdded is null");
+            this.rowsAdded = requireNonNull(rowsAdded, "rowsAdded is null");
+            this.serde = requireNonNull(serde, "serde is null");
+            this.lifespan = requireNonNull(lifespan, "lifespan is null");
+        }
+
+        public void resetPositionCount()
+        {
+            positionCount = 0;
+        }
+
+        public void appendPosition(int position)
+        {
+            positions[positionCount++] = position;
+        }
+
+        public void appendRows(DecodedBlockNode[] decodedBlocks, int fixedWidthRowSize, List<Integer> variableWidthChannels, OutputBuffer outputBuffer)
+        {
+            checkArgument(decodedBlocks.length == channelCount, "Unexpected number of blocks");
+
+            if (positionCount == 0) {
+                return;
+            }
+
+            if (channelCount == 0) {
+                bufferedRowCount += positionCount;
+                return;
+            }
+
+            initializeBlockEncodingBuffers(decodedBlocks);
+
+            for (int i = 0; i < channelCount; i++) {
+                blockEncodingBuffers[i].setupDecodedBlocksAndPositions(decodedBlocks[i], positions, positionCount);
+            }
+
+            calculateRowSizes(fixedWidthRowSize, variableWidthChannels);
+
+            int offset = 0;
+            do {
+                int batchSize = calculateNextBatchSize(fixedWidthRowSize, variableWidthChannels, offset);
+
+                for (int i = 0; i < channelCount; i++) {
+                    blockEncodingBuffers[i].setNextBatch(offset, batchSize);
+                    blockEncodingBuffers[i].copyValues();
+                }
+
+                bufferedRowCount += batchSize;
+                offset += batchSize;
+
+                if (bufferFull) {
+                    flush(outputBuffer);
+
+                    bufferFull = false;
+                }
+            }
+            while (offset < positionCount);
+        }
+
+        public long getSizeInBytes()
+        {
+            long buffersSizeInBytes = 0;
+            if (blockEncodingBuffers != null) {
+                for (int i = 0; i < channelCount; i++) {
+                    buffersSizeInBytes += blockEncodingBuffers[i].getSizeInBytes();
+                }
+            }
+
+            // Add up sizes for positions and rowSizes arrays
+            return SIZE_OF_INT * 2 * positionCount + buffersSizeInBytes;
+        }
+
+        public long getRetainedSizeInBytes()
+        {
+            long buffersSizeInBytes = 0;
```

Comment:
- buffersSizeInBytes -> size for brevity
- consider initializing this variable with a non-zero value

```
            long size = INSTANCE_SIZE + sizeOf(positions) + sizeOf(rowSizes);
            if (blockEncodingBuffers != null) {
                for (int i = 0; i < channelCount; i++) {
                    size += blockEncodingBuffers[i].getRetainedSizeInBytes();
                }
            }

            return size;
```